pandas/core/groupby/generic.py

```diff
@@ -1754,10 +1757,27 @@ def value_counts(
                     level=index_level, sort_remaining=False
                 )
 
-            if not self.as_index:
+            if self.as_index:
+                return result.__finalize__(self.obj, method="value_counts")
+            else:
                 # Convert to frame
-                result = result.reset_index(name="proportion" if normalize else "count")
-            return result.__finalize__(self.obj, method="value_counts")
+                name = "proportion" if normalize else "count"
+                columns = result.index.names
+                if name in columns:
+                    raise ValueError(
+                        f"Column label '{name}' is duplicate of result column"
+                    )
+                columns = com.fill_missing_names(columns)
+                result_frame = DataFrame()
+                for i, column in enumerate(columns):
+                    level_values = result.index.get_level_values(i)._values
+                    if level_values.dtype == np.object_:
+                        level_values = lib.maybe_convert_objects(
+                            cast(np.ndarray, level_values)
+                        )
+                    result_frame.insert(i, column, level_values, allow_duplicates=True)
```

Comment:
this almost O(n^2) inefficient.
so what you need to do is
- if we don't have duplicates in the columns names, concat and are done
- if we do, then a) should this really be the answer? why are we not raising?
- if we actually do allow this then you simply change the column names to a range index, concat, then set them again.


pls don't conflate this issue with reset_index or allow_duplicates, they are completely orthogonal.

this is not going to move forward this keeps re-inventing the wheel.

Reply:
Not O(n^2)! But O(n), and of course I would rather not be itererating through an axis. That is the whole point of Pandas!

concat does not work, as I explained many times. The best I could do was:

```
result_frame = index.set_names(range(len(columns))).to_frame()
result_frame = concat([result_frame, result], axis=1).reset_index(drop=True).set_axis(columns + [name], axis=1)
```

which fails 6 of my tests due to bool/object problems in MultiIndex. These will probably be fixed by #45061, but I see that has been deferred until 1.5.

Meanwhile, I employed your column renaming suggestion. It is now all green (apart from the usual s3 problem)